sync kvm patch for loongarch

lixianglai opened the pull request:
## Summary by Sourcery

This pull request synchronizes the KVM patch for the LoongArch architecture. It introduces support for performance monitoring units (PMU), LBT, stolen time, and IPI. It also includes changes to the MMU, timer, and other core components to improve virtualization performance and stability.

New Features:
- Adds support for performance monitoring units (PMU) to allow guest VMs to monitor their performance.
- Introduces support for LBT (likely Load-and-Branch Tracking) to improve the performance of nested virtualization.
- Implements stolen time accounting to provide more accurate timekeeping in guest VMs.
- Adds support for Inter-Processor Interrupts (IPI) to enable communication between virtual CPUs.
- Adds support for virtualized EXTIOI (External Interrupt Controller)

Commit:
LoongArch: Fix virtual machine startup error

Upstream: no

Signed-off-by: liuyun <liuyun@loongson.cn>
Signed-off-by: maobibo <maobibo@loongson.cn>
Signed-off-by: Ming Wang <wangming01@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Remove unnecessary CSR register saving during enter guest

commit e2f77af9ec6e ("LoongArch: KVM: Remove unnecessary CSR register saving during enter guest")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Some CSR registers like CRMD/PRMD are saved during enter VM mode now.
However they are not restored for actual use, so saving for these CSR
registers can be removed.

Reviewed-by: Tianrui Zhao <zhaotianrui@loongson.cn>
Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

seehearfeel committed:
LoongArch: Add ORC stack unwinder support

commit cb8a2ef0848c ("LoongArch: Add ORC stack unwinder support")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

The kernel CONFIG_UNWINDER_ORC option enables the ORC unwinder, which is
similar in concept to a DWARF unwinder. The difference is that the format
of the ORC data is much simpler than DWARF, which in turn allows the ORC
unwinder to be much simpler and faster.

The ORC data consists of unwind tables which are generated by objtool.
After analyzing all the code paths of a .o file, it determines information
about the stack state at each instruction address in the file and outputs
that information to the .orc_unwind and .orc_unwind_ip sections.

The per-object ORC sections are combined at link time and are sorted and
post-processed at boot time. The unwinder uses the resulting data to
correlate instruction addresses with their stack states at run time.

Most of the logic are similar with x86, in order to get ra info before ra
is saved into stack, add ra_reg and ra_offset into orc_entry. At the same
time, modify some arch-specific code to silence the objtool warnings.

Change-Id: I01f9e01fb0bc659e5c430c1871b5093d6ee684ba
Co-developed-by: Jinyang He <hejinyang@loongson.cn>
Signed-off-by: Jinyang He <hejinyang@loongson.cn>
Co-developed-by: Youling Tang <tangyouling@loongson.cn>
Signed-off-by: Youling Tang <tangyouling@loongson.cn>
Signed-off-by: Tiezhu Yang <yangtiezhu@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: irqchip/loongson-eiointc: Add virt extension support

Upstream: no

With virt eiointc, interrupt can be routed to 256 vcpus

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Add mmio trace events support

commit 4bf3b972cad4 ("LoongArch: KVM: Add mmio trace events support")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Add mmio trace events support, currently generic mmio events
KVM_TRACE_MMIO_WRITE/xxx_READ/xx_READ_UNSATISFIED are added here.

Also vcpu id field is added for all kvm trace events, since perf
KVM tool parses vcpu id information for kvm entry event.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Add vcpu mapping from physical cpuid

commit a21f6257348f ("LoongArch: KVM: Add vcpu mapping from physical cpuid")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Physical CPUID is used for interrupt routing for irqchips such as ipi,
msgint and eiointc interrupt controllers. Physical CPUID is stored at
the CSR register LOONGARCH_CSR_CPUID, it can not be changed once vcpu
is created and the physical CPUIDs of two vcpus cannot be the same.

Different irqchips have different size declaration about physical CPUID,
the max CPUID value for CSR LOONGARCH_CSR_CPUID on Loongson-3A5000 is
512, the max CPUID supported by IPI hardware is 1024, while for eiointc
irqchip is 256, and for msgint irqchip is 65536.

The smallest value from all interrupt controllers is selected now, and
the max cpuid size is defines as 256 by KVM which comes from the eiointc
irqchip.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Add PV IPI support on host side

Upstream: no

On LoongArch system, IPI hw uses iocsr registers. There are one iocsr
register access on IPI sending, and two iocsr access on IPI receiving
for the IPI interrupt handler. In VM mode all iocsr accessing will cause
VM to trap into hypervisor. So with one IPI hw notification there will
be three times of trap.

In this patch PV IPI is added for VM, hypercall instruction is used for
IPI sender, and hypervisor will inject an SWI to the destination vcpu.
During the SWI interrupt handler, only CSR.ESTAT register is written to
clear irq. CSR.ESTAT register access will not trap into hypervisor, so
with PV IPI supported, there is one trap with IPI sender, and no trap
with IPI receiver, there is only one trap with IPI notification.

Also this patch adds IPI multicast support, the method is similar with
x86. With IPI multicast support, IPI notification can be sent to at
most 128 vcpus at one time. It greatly reduces the times of trapping
into hypervisor.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Add iocsr and mmio bus simulation in kernel

Upstream: no

Add iocsr and mmio memory read and write simulation to the kernel.
When the VM accesses the device address space through iocsr
instructions or mmio, it does not need to return to the qemu
user mode but directly completes the access in the kernel mode.

Signed-off-by: Tianrui Zhao <zhaotianrui@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Add IPI device support

Upstream: no

Added device model for IPI interrupt controller,
implemented basic create destroy interface,
and registered device model to kvm device table.

Signed-off-by: Tianrui Zhao <zhaotianrui@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Add EXTIOI device support

Upstream: no

Added device model for EXTIOI interrupt controller,
implemented basic create destroy interface,
and registered device model to kvm device table.

Signed-off-by: Tianrui Zhao <zhaotianrui@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Add PCHPIC device support

Upstream: no

Added device model for PCHPIC interrupt controller,
implemented basic create destroy interface,
and registered device model to kvm device table.

Signed-off-by: Tianrui Zhao <zhaotianrui@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Add irqfd support

Upstream: no

Enable the KVM_IRQ_ROUTING KVM_IRQCHIP KVM_MSI configuration item,
increase the KVM_CAP_IRQCHIP capability, and implement the query
interface of the kernel irqchip.

Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
KVM: Discard zero mask with function kvm_dirty_ring_reset

commit 05086585ea4f ("KVM: Discard zero mask with function kvm_dirty_ring_reset")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Function kvm_reset_dirty_gfn may be called with parameters cur_slot /
cur_offset / mask are all zero, it does not represent real dirty page.
It is not necessary to clear dirty page in this condition. Also return
value of macro __fls() is undefined if mask is zero which is called in
funciton kvm_reset_dirty_gfn(). Here just return.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Message-ID: <20240613122803.1031511-1-maobibo@loongson.cn>
[Move the conditional inside kvm_reset_dirty_gfn; suggested by
 Sean Christopherson. - Paolo]
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Add IPI read and write function

Upstream: no

Implementation of IPI interrupt controller address
space read and write function simulation.

Signed-off-by: Min Zhou <zhoumin@loongson.cn>
Signed-off-by: Tianrui Zhao <zhaotianrui@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Add IPI user mode read and write function

Upstream: no

Implements the communication interface between the user mode
program and the kernel in IPI interrupt control simulation,
which is used to obtain or send the simulation data of the
interrupt controller in the user mode process, and is used
in VM migration or VM saving and restoration.

Signed-off-by: Min Zhou <zhoumin@loongson.cn>
Signed-off-by: Tianrui Zhao <zhaotianrui@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Add EXTIOI read and write functions

Upstream: no

Implementation of EXTIOI interrupt controller address
space read and write function simulation.

Signed-off-by: Tianrui Zhao <zhaotianrui@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Add EXTIOI user mode read and write functions

Upstream: no

Implements the communication interface between the user mode
program and the kernel in EXTIOI interrupt control simulation,
which is used to obtain or send the simulation data of the
interrupt controller in the user mode process, and is used
in VM migration or VM saving and restoration.

Signed-off-by: Tianrui Zhao <zhaotianrui@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Add PCHPIC read and write functions

Upstream: no

Implementation of IPI interrupt controller address
space read and write function simulation.

Implement interrupt injection interface under loongarch.

Signed-off-by: Tianrui Zhao <zhaotianrui@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Add PCHPIC user mode read and write functions

Upstream: no

Implements the communication interface between the user mode
program and the kernel in PCHPIC interrupt control simulation,
which is used to obtain or send the simulation data of the
interrupt controller in the user mode process, and is used
in VM migration or VM saving and restoration.

Signed-off-by: Tianrui Zhao <zhaotianrui@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Delay secondary mmu tlb flush until guest entry

commit 846f1299b2e2 ("LoongArch: KVM: Delay secondary mmu tlb flush until guest entry")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

With hardware assisted virtualization, there are two level HW mmu, one
is GVA to GPA mapping, the other is GPA to HPA mapping which is called
secondary mmu in generic. If there is page fault for secondary mmu,
there needs tlb flush operation indexed with fault GPA address and VMID.
VMID is stored at register CSR.GSTAT and will be reload or recalculated
before guest entry.

Currently CSR.GSTAT is not saved and restored during VCPU context
switch, instead it is recalculated during guest entry. So CSR.GSTAT is
effective only when a VCPU runs in guest mode, however it may not be
effective if the VCPU exits to host mode. Since register CSR.GSTAT may
be stale, it may records the VMID of the last schedule-out VCPU, rather
than the current VCPU.

Function kvm_flush_tlb_gpa() should be called with its real VMID, so
here move it to the guest entrance. Also an arch-specific request id
KVM_REQ_TLB_FLUSH_GPA is added to flush tlb for secondary mmu, and it
can be optimized if VMID is updated, since all guest tlb entries will
be invalid if VMID is updated.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Sync pending interrupt when getting ESTAT from user mode

commit 2584b0859859 ("LoongArch: KVM: Sync pending interrupt when getting ESTAT from user mode")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Currently interrupts are posted and cleared with the asynchronous mode,
meanwhile they are saved in SW state vcpu::arch::irq_pending and vcpu::
arch::irq_clear. When vcpu is ready to run, pending interrupt is written
back to CSR.ESTAT register from SW state vcpu::arch::irq_pending at the
guest entrance.

During VM migration stage, vcpu is put into stopped state, however
pending interrupts are not synced to CSR.ESTAT register. So there will
be interrupt lost when VCPU is migrated to another host machines.

Here in this patch when ESTAT CSR register is read from VMM user mode,
pending interrupts are synchronized to ESTAT also. So that VMM can get
correct pending interrupts.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

Commit:
LoongArch: KVM: always make pte young in page map's fast path

commit d808899636ea ("LoongArch: KVM: always make pte young in page map's fast path")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

It seems redundant to check if pte is young before the call to
kvm_pte_mkyoung() in kvm_map_page_fast(). Just remove the check.

Reviewed-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Jia Qingtong <jiaqingtong97@gmail.com>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Add memory barrier before update pmd entry

commit 21d29364e1d1 ("LoongArch: KVM: Add memory barrier before update pmd entry")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

When updating pmd entry such as allocating new pmd page or splitting
huge page into normal page, it is necessary to firstly update all pte
entries, and then update pmd entry.

It is weak order with LoongArch system, there will be problem if other
VCPUs see pmd update firstly while ptes are not updated. Here smp_wmb()
is added to assure this.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Add dirty bitmap initially all set support

commit e31e55145f76 (""LoongArch: KVM: Add dirty bitmap initially all set support)
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Add KVM_DIRTY_LOG_INITIALLY_SET support on LoongArch system, this
feature comes from other architectures like x86 and arm64.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Mark page accessed and dirty with page ref added

commit dfd5f11227e6 ("LoongArch: KVM: Mark page accessed and dirty with page ref added")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Function kvm_map_page_fast() is fast path of secondary mmu page fault
flow, pfn is parsed from secondary mmu page table walker. However the
corresponding page reference is not added, it is dangerious to access
page out of mmu_lock.

Here page ref is added inside mmu_lock, function kvm_set_pfn_accessed()
and kvm_set_pfn_dirty() is called with page ref added, so that the page
will not be freed by others.

Also kvm_set_pfn_accessed() is removed here since it is called in the
following function kvm_release_pfn_clean().

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Add PV steal time support in host side

commit 40e12dbc794b ("LoongArch: KVM: Add PV steal time support in host side")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Add ParaVirt steal time feature in host side, VM can search supported
features provided by KVM hypervisor, a feature KVM_FEATURE_STEAL_TIME
is added here. Like x86, steal time structure is saved in guest memory,
one hypercall function KVM_HCALL_FUNC_NOTIFY is added to notify KVM to
enable this feature.

One CPU attr ioctl command KVM_LOONGARCH_VCPU_PVTIME_CTRL is added to
save and restore the base address of steal time structure when a VM is
migrated.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Add PV steal time support in guest side

commit 00c8fa6c61c5 (""LoongArch: KVM: Add PV steal time support in guest side)
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Per-cpu struct kvm_steal_time is added here, its size is 64 bytes and
also defined as 64 bytes, so that the whole structure is in one physical
page.

When a VCPU is online, function pv_enable_steal_time() is called. This
function will pass guest physical address of struct kvm_steal_time and
tells hypervisor to enable steal time. When a vcpu is offline, physical
address is set as 0 and tells hypervisor to disable steal time.

Here is an output of vmstat on guest when there is workload on both host
and guest. It shows steal time stat information.

procs -----------memory---------- -----io---- -system-- ------cpu-----
 r  b   swpd   free  inact active   bi    bo   in   cs us sy id wa st
15  1      0 7583616 184112  72208    20    0  162   52 31  6 43  0 20
17  0      0 7583616 184704  72192    0     0 6318 6885  5 60  8  5 22
16  0      0 7583616 185392  72144    0     0 1766 1081  0 49  0  1 50
16  0      0 7583616 184816  72304    0     0 6300 6166  4 62 12  2 20
18  0      0 7583632 184480  72240    0     0 2814 1754  2 58  4  1 35

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Discard dirty page tracking on readonly memslot

commit 1706dcacd354 ("LoongArch: KVM: Discard dirty page tracking on readonly memslot")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

For readonly memslot such as UEFI BIOS or UEFI var space, guest cannot
write this memory space directly. So it is not necessary to track dirty
pages for readonly memslot. Here we make such optimization in function
kvm_arch_commit_memory_region().

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
perf kvm: Add kvm-stat for loongarch64

commit 0063265ee183 ("perf kvm: Add kvm-stat for loongarch64")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Add support for 'perf kvm stat' on loongarch64 platform, now only kvm
exit event is supported.

Here is example output about "perf kvm --host stat report" command

   Event name   Samples   Sample%     Time (ns)   Time%   Mean Time (ns)
    Mem Store     83969    51.00%     625697070   8.00%             7451
     Mem Read     37641    22.00%     112485730   1.00%             2988
    Interrupt     15542     9.00%      20620190   0.00%             1326
        IOCSR     15207     9.00%      94296190   1.00%             6200
    Hypercall      4873     2.00%      12265280   0.00%             2516
         Idle      3713     2.00%    6322055860  87.00%          1702681
          FPU      1819     1.00%       2750300   0.00%             1511
   Inst Fetch       502     0.00%       1341740   0.00%             2672
   Mem Modify       324     0.00%        602240   0.00%             1858
       CPUCFG        55     0.00%         77610   0.00%             1411
          CSR        12     0.00%         19690   0.00%             1640
         LASX         3     0.00%          4870   0.00%             1623
          LSX         2     0.00%          2100   0.00%             1050

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

Commit:
LoongArch: KVM: Remove undefined a6 argument comment for kvm_hypercall()

commit bcd4e6fa8cf0 ("LoongArch: KVM: Remove undefined a6 argument comment for kvm_hypercall()")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

The kvm_hypercall() set for LoongArch is limited to a1-a5. So the
mention of a6 in the comment is undefined that needs to be rectified.

Reviewed-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Wentao Guan <guanwentao@uniontech.com>
Signed-off-by: Dandan Zhang <zhangdandan@uniontech.com>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

mstsirkin committed:
virtio: break and reset virtio devices on device_shutdown()

commit 8bd2fa086a04 ("LoongArch: Add ifdefs to fix LSX and LASX related warnings")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Hongyu reported a hang on kexec in a VM. QEMU reported invalid memory
accesses during the hang.

Invalid read at addr 0x102877002, size 2,

region '(null)', reason: rejected

Invalid write at addr 0x102877A44, size 2,

region '(null)', reason: rejected

...

It was traced down to virtio-console. Kexec works fine if virtio-console
is not in use.

The issue is that virtio-console continues to write to the MMIO even after
underlying virtio-pci device is reset.

Additionally, Eric noticed that IOMMUs are reset before devices, if
devices are not reset on shutdown they continue to poke at guest memory
and get errors from the IOMMU. Some devices get wedged then.

The problem can be solved by breaking all virtio devices on virtio
bus shutdown, then resetting them.

Tested-by: Eric Auger <eric.auger@redhat.com>
Acked-by: Jason Wang <jasowang@redhat.com>
Signed-off-by: Michael S. Tsirkin <mst@redhat.com>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>
Change-Id: I5f1f050c48cd74fc0b0b759cc7287aa939845b32

lixianglai committed:
LoongArch: drivers/iommu: add iommu support

Upstream: no

Added iommu support for loongarch

Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

seehearfeel committed:
LoongArch: Add ifdefs to fix LSX and LASX related warnings

commit 80376323e2b6 ("LoongArch: Add ifdefs to fix LSX and LASX related warnings")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

There exist some warnings when building kernel if CONFIG_CPU_HAS_LBT is
set but CONFIG_CPU_HAS_LSX and CONFIG_CPU_HAS_LASX are not set. In this
case, there are no definitions of _restore_lsx & _restore_lasx and there
are also no definitions of kvm_restore_lsx & kvm_restore_lasx in fpu.S
and switch.S respectively, just add some ifdefs to fix these warnings.

  AS      arch/loongarch/kernel/fpu.o
arch/loongarch/kernel/fpu.o: warning: objtool: unexpected relocation symbol type in .rela.discard.func_stack_frame_non_standard: 0
arch/loongarch/kernel/fpu.o: warning: objtool: unexpected relocation symbol type in .rela.discard.func_stack_frame_non_standard: 0

  AS [M]  arch/loongarch/kvm/switch.o
arch/loongarch/kvm/switch.o: warning: objtool: unexpected relocation symbol type in .rela.discard.func_stack_frame_non_standard: 0
arch/loongarch/kvm/switch.o: warning: objtool: unexpected relocation symbol type in .rela.discard.func_stack_frame_non_standard: 0

  MODPOST Module.symvers
ERROR: modpost: "kvm_restore_lsx" [arch/loongarch/kvm/kvm.ko] undefined!
ERROR: modpost: "kvm_restore_lasx" [arch/loongarch/kvm/kvm.ko] undefined!

Change-Id: I946d251eef85aed95ddc93f38c0cc6c7fa725dc7
Cc: stable@vger.kernel.org # 6.9+
Fixes: cb8a2ef0848c ("LoongArch: Add ORC stack unwinder support")
Reported-by: kernel test robot <lkp@intel.com>
Closes: https://lore.kernel.org/oe-kbuild-all/202408120955.qls5oNQY-lkp@intel.com/
Signed-off-by: Tiezhu Yang <yangtiezhu@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Invalidate guest steal time address on vCPU reset

commit 0c9fa3e92629 ("LoongArch: KVM: Invalidate guest steal time address on vCPU reset")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

If ParaVirt steal time feature is enabled, there is a percpu gpa address
passed from guest vCPU and host modifies guest memory space with this gpa
address. When vCPU is reset normally, it will notify host and invalidate
gpa address.

However if VM is crashed and VMM reboots VM forcely, the vCPU reboot
notification callback will not be called in VM. Host needs invalidate
the gpa address, else host will modify guest memory during VM reboots.
Here it is invalidated from the vCPU KVM_REG_LOONGARCH_VCPU_RESET ioctl
interface.

Also funciton kvm_reset_timer() is removed at vCPU reset stage, since SW
emulated timer is only used in vCPU block state. When a vCPU is removed
from the block waiting queue, kvm_restore_timer() is called and SW timer
is cancelled. And the timer register is also cleared at VMM when a vCPU
is reset.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: Revert qspinlock to test-and-set simple lock on VM

commit 9d4885312bef ("LoongArch: Revert qspinlock to test-and-set simple lock on VM")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Similar with x86, when VM is detected, revert to a simple test-and-set
lock to avoid the horrors of queue preemption.

Tested on 3C5000 Dual-way machine with 32 cores and 2 numa nodes,
test case is kcbench on kernel mainline 6.10, the detailed command is
"kcbench --src /root/src/linux"

Performance on host machine
                      kernel compile time       performance impact
   Original           150.29 seconds
   With patch         150.19 seconds            almost no impact

Performance on virtual machine:
1. 1 VM with 32 vCPUs and 2 numa node, numa node pinned
                      kernel compile time       performance impact
   Original           170.87 seconds
   With patch         171.73 seconds            almost no impact

2. 2 VMs, each VM with 32 vCPUs and 2 numa node, numa node pinned
                      kernel compile time       performance impact
   Original           2362.04 seconds
   With patch         354.73  seconds            +565%

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Add Binary Translation extension support

commit 59d8372bfc46 ("LoongArch: KVM: Add Binary Translation extension support")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Loongson Binary Translation (LBT) is used to accelerate binary translation,
which contains 4 scratch registers (scr0 to scr3), x86/ARM eflags (eflags)
and x87 fpu stack pointer (ftop).

Like FPU extension, here a lazy enabling method is used for LBT. the LBT
context is saved/restored on the vcpu context switch path.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Add vm migration support for LBT registers

commit 3462e7fba862 ("LoongArch: KVM: Add vm migration support for LBT registers")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Every vcpu has separate LBT registers. And there are four scr registers,
one flags and ftop register for LBT extension. When VM migrates, VMM
needs to get LBT registers for every vcpu.

Here macro KVM_REG_LOONGARCH_LBT is added for new vcpu lbt register type,
the following macro is added to get/put LBT registers.
  KVM_REG_LOONGARCH_LBT_SCR0
  KVM_REG_LOONGARCH_LBT_SCR1
  KVM_REG_LOONGARCH_LBT_SCR2
  KVM_REG_LOONGARCH_LBT_SCR3
  KVM_REG_LOONGARCH_LBT_EFLAGS
  KVM_REG_LOONGARCH_LBT_FTOP

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Enable paravirt feature control from VMM

commit 14d76d12c574 ("LoongArch: KVM: Enable paravirt feature control from VMM")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Export kernel paravirt features to user space, so that VMM can control
each single paravirt feature. By default paravirt features will be the
same with kvm supported features if VMM does not set it.

Also a new feature KVM_FEATURE_VIRT_EXTIOI is added which can be set
from user space. This feature indicates that the virt EIOINTC can route
interrupts to 256 vCPUs, rather than 4 vCPUs like with real HW.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: KVM: Implement function kvm_para_has_feature()

commit cbc93f3256f3 ("LoongArch: KVM: Implement function kvm_para_has_feature()")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Implement function kvm_para_has_feature() to detect supported paravirt
features. It can be used by device driver to detect and enable paravirt
features, such as the EIOINTC irqchip driver is able to detect feature
KVM_FEATURE_VIRT_EXTIOI and do some optimization.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: Fix the issue of code inconsistency after adjusting the patch sequence

Upstream: no

Due to the inconsistency between the previously merged patch
and the upstream merged patch, as well as the adjustment of
the patch order, the virtual machine code differs from the
code of other domestic communities.

In order to keep the virtualization code of loongarch in
various communities consistent, the current patch is now
submitted to modify and adjust the differences.

The adjustment involves the following patches:
6ad23dd1df32 ("LoongArch: KVM: Add software breakpoint support")
6bc0120a72a0 ("LoongArch: KVM: Add hypercall instruction emulation")
659e2213eaa2 ("LoongArch: KVM: Add PV IPI support on guest side")
4e04e9836386 ("LoongArch: KVM: Implement kvm mmu operations")

Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: arch/loongarch/kvm: Add reset function for extioi

Upstream: no

When the virtual machine is restarted, the data in extioi is not zeroed,
and there is a residual set interrupt bit, resulting in a hang

Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

wang-whl committed:
LoongArch: eiointc:Fix ext irq route error.

Upstream: no

Signed-off-by: wanghongliang <wanghongliang@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

bibo-mao committed:
LoongArch: LSVZ: Clear LLBCTL if secondary mmu mapping is changed

Upstream: no

Register LLBCTL is separated CSR register from host, host exception
eret instruction will clear host LLBCTL CSR register, guest
exception will clear guest LLBCTL CSR register.

VCPU0 atomic64_fetch_add_unless      VCPU1 atomic64_fetch_add_unless
  0:   ll.d    %[p],  %[c]
       beq     %[p],  %[u], 1f
Here secondary mmu mapping is changed, hpa page is is replaced
with new page. And VCPU1 executed atomic instruction on new
page.
                                     0:   ll.d    %[p],  %[c]
                                          beq     %[p],  %[u], 1f
                                          add.d   %[rc], %[p], %[a]
                                          sc.d    %[rc], %[c]
       add.d   %[rc], %[p], %[a]
       sc.d    %[rc], %[c]
LLBCTL is on and it represents the memory is not modified, sc.d
will modify the memory directly.

Here clear guest LLBCTL_WCLLB register when mapping is the changed.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: enable ptw for kvm

Upstream: no

Add the ptw feature bit to cpucfg

Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Add reset function for irqchip

Upstream: no

We need to empty the data in irqchip when the VM restarts.

Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

opsiff commented on arch/loongarch/include/asm/pgtable.h:
```diff
@@ -329,12 +329,13 @@ static inline void set_pte(pte_t *ptep, pte_t pteval)
 {
-	pte_t pte = ptep_get(ptep);
+	/* Preserve global status for the pair */
+	pte_t pte = *(ptep);
```
It seems different with upstream commit,why?
LoongArch: Set initial pte entry with PAGE_GLOBAL for kernel space
Link: https://web.git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/arch/loongarch/include/asm/pgtable.h?h=v6.14-rc4&id=d2f8671045b41871053dedaf3035a06ad53d2736

lixianglai replied:
这个确实有些问题，deepin的代码比较，打上的patch相对老一些，在解决冲突的时候没有注意到这一点，应该以上游的内核代码为基准解决冲突，我会修复它

opsiff commented on drivers/iommu/dma-iommu.c:
```diff
@@ -1640,7 +1640,7 @@ static size_t iommu_dma_max_mapping_size(struct device *dev)
 }
 
-static const struct dma_map_ops iommu_dma_ops = {
+static const struct dma_map_ops iommu_dmafops = {
```
能否不修改此处名称? 似乎没必要修改。

lixianglai replied:
这个确实有问题，我会修复它的

opsiff commented on arch/loongarch/configs/loongson3_defconfig:
```diff
@@ -2204,3 +2204,6 @@ CONFIG_RCU_CPU_STALL_TIMEOUT=60
 # CONFIG_STRICT_DEVMEM is not set
 # CONFIG_RUNTIME_TESTING_MENU is not set
 CONFIG_UNWINDER_ORC=y
+CONFIG_CMDLINE_EXTEND=y
```
需要一并修改 deepin_loongarch_desktop_defconfig

lixianglai replied:
好的

opsiff commented on arch/loongarch/include/asm/kvm_ipi.h:
```diff
@@ -0,0 +1,36 @@
+ * Copyright (C) 2024 Loongson Technology Corporation Limited
+ */
+
+#ifndef __LS3A_KVM_IPI_H
```
这个文件夹下其他地方定义类似于 __ASM_LOONGARCH_KVM_CSR_H__
文件名宏是否考虑统一命名?

lixianglai replied:
已修改

Avenger-285714 commented on arch/loongarch/kvm/intc/extioi.c:
```diff
@@ -0,0 +1,956 @@
+	case 8:
+		ret = loongarch_extioi_writel(vcpu, extioi, addr, len, val);
+		break;
+	default:
```
default时ret未赋值。

Avenger-285714 commented on arch/loongarch/kvm/intc/extioi.c:
```diff
@@ -0,0 +1,956 @@
+	}
+
+	loongarch_ext_irq_unlock(extioi, flags);
+
```
多余的空行？

Avenger-285714 commented on arch/loongarch/kvm/intc/extioi.c:
```diff
@@ -0,0 +1,956 @@
+	case 8:
+		ret = loongarch_extioi_readl(vcpu, extioi, addr, len, val);
+		break;
+	default:
```
default时ret未赋值。

opsiff commented on drivers/iommu/loongarch_iommu.c:
```diff
@@ -0,0 +1,1728 @@
+__setup("loongarch_iommu=", la_iommu_setup);
+
+static const struct pci_device_id loongson_iommu_pci_tbl[] = {
+	{ PCI_DEVICE(0x14, 0x3c0f) },
```
此处是否使用PCI_VENDOR_ID_LOONGSON?

opsiff commented on drivers/iommu/loongarch_iommu.c:
```diff
@@ -0,0 +1,1728 @@
+			iommu = kzalloc(sizeof(struct loongarch_iommu),
+					GFP_KERNEL);
+			if (iommu == NULL)
+				return -ENOMEM;
```
如果初始化多个p 是否此处错误处理应该释放前面所有的p?

opsiff commented on drivers/iommu/loongarch_iommu.c:
```diff
@@ -0,0 +1,1728 @@
+				return ret;
+		}
+
+		if (h->length == 0)
```
此处if (h->type == la_iommu_target_ivhd_type) 是否需要放在if (h->type == la_iommu_target_ivhd_type)前面?

lixianglai replied:
??

opsiff replied:
> ??
回复错了 意思是在下面的逻辑中先判断if (h->length == 0)
为什么在这里先判断if (h->type == la_iommu_target_ivhd_type)
```
+static int __init init_iommu_all(struct acpi_table_header *table)
+{
+	u8 *p = (u8 *)table, *end = (u8 *)table;
+	struct ivhd_header *h;
+	struct loongarch_iommu *iommu;
+	int ret;
+
+	end += table->length;
+	p += IVRS_HEADER_LENGTH;
+
+	while (p < end) {
+		h = (struct ivhd_header *)p;
+
+		if (h->length == 0)
+			break;
+
+		if (*p == la_iommu_target_ivhd_type) {
+
+			pr_info("device: %02x:%02x.%01x seg: %d\n",
+				    PCI_BUS_NUM(h->devid), PCI_SLOT(h->devid),
+				    PCI_FUNC(h->devid), h->pci_seg);
+
+			iommu = kzalloc(sizeof(struct loongarch_iommu),
+					GFP_KERNEL);
+			if (iommu == NULL)
+				return -ENOMEM;
+
+			ret = init_iommu_one(iommu, h);
+			if (ret) {
+				kfree(iommu);
+				pr_info("%s init iommu failed\n", __func__);
+				return ret;
+			}
+		}
+		p += h->length;
```

lixianglai replied:
从逻辑上来说，应该先判断len==0,如果成立就不再解析内容了，这里应该是没有问题的，我去改一下上面的判断逻辑

opsiff commented on drivers/iommu/loongarch_iommu.c:
```diff
@@ -0,0 +1,1728 @@
+		if (h->length == 0)
+			break;
+
+		if (*p == la_iommu_target_ivhd_type) {
```
另外一个地方用的 if (h->type == la_iommu_target_ivhd_type)  能不能统一?

lixianglai committed:
LoongArch: KVM: Reload guest CSR registers after S4

Upstream: no

On host HW guest CSR registers are lost after suspend and resume
operation. Since last_vcpu of boot CPU still records latest vCPU pointer
so that guest CSR register skips to reload when boot CPU resumes and
vCPU is scheduled.

Here last_vcpu is cleared so that guest CSR registers will reload from
scheduled vCPU context after suspend and resume.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Add interrupt checking with Loongson AVEC

Upstream: no

There is newly added macro INT_AVEC with CSR ESTAT register, which is
bit 14 used for Loongson AVEC support. AVEC interrupt status bit 14 is
supported with macro CSR_ESTAT_IS, here replace hardcoded value 0x1fff
with macro CSR_ESTAT_IS so that AVEC interrupt status is supported by
KVM also.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Select huge page only if secondary mmu supports it

commit 038f365107ed ("LoongArch: KVM: Select huge page only if secondary mmu supports it")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Currently page level selection about secondary mmu depends on memory
slot and page level about host mmu. There will be problems if page level
of secondary mmu is zero already. Huge page cannot be selected if there is
normal page mapped in secondary mmu already, since it is not supported to
merge normal pages into huge pages now.

So page level selection should depend on the following three conditions.
 1. Memslot is aligned for huge page and vm is not migrating.
 2. Page level of host mmu is also huge page.
 3. Page level of secondary mmu is suituable for huge page.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Add VM feature detection function

commit 3e526eda2362 ("LoongArch: KVM: Add VM feature detection function")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Loongson SIMD Extension (LSX), Loongson Advanced SIMD Extension (LASX)
and Loongson Binary Translation (LBT) features are defined in register
CPUCFG2. Two kinds of LSX/LASX/LBT feature detection are added here, one
is VCPU feature, and the other is VM feature. VCPU feature dection can
only work with VCPU thread itself, and requires VCPU thread is created
already. So LSX/LASX/LBT feature detection for VM is added also, it can
be done even if VM is not created, and also can be done by any threads
besides VCPU threads.

Here ioctl command KVM_HAS_DEVICE_ATTR is added for VM, and macro
KVM_LOONGARCH_VM_FEAT_CTRL is added to check supported feature. And
five sub-features relative with LSX/LASX/LBT are added as following:
 KVM_LOONGARCH_VM_FEAT_LSX
 KVM_LOONGARCH_VM_FEAT_LASX
 KVM_LOONGARCH_VM_FEAT_X86BT
 KVM_LOONGARCH_VM_FEAT_ARMBT
 KVM_LOONGARCH_VM_FEAT_MIPSBT

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Add PMU support for guest

commit e47d24922e1a ("LoongArch: KVM: Add PMU support for guest")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

On LoongArch, the host and guest have their own PMU CSRs registers and
they share PMU hardware resources. A set of PMU CSRs consists of a CTRL
register and a CNTR register. We can set which PMU CSRs are used by the
guest by writing to the GCFG register [24:26] bits.

On KVM side:
- Save the host PMU CSRs into structure kvm_context.
- If the host supports the PMU feature.
  - When entering guest mode, save the host PMU CSRs and restore the guest PMU CSRs.
  - When exiting guest mode, save the guest PMU CSRs and restore the host PMU CSRs.

Reviewed-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Song Gao <gaosong@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Add cpucfg area for kvm hypervisor

commit 8dc1a4832dd4 ("LoongArch: KVM: Add cpucfg area for kvm hypervisor")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

Instruction cpucfg can be used to get processor features. And there
is a trap exception when it is executed in VM mode, and also it can be
used to provide cpu features to VM. On real hardware cpucfg area 0 - 20
is used by now. Here one specified area 0x40000000 -- 0x400000ff is used
for KVM hypervisor to provide PV features, and the area can be extended
for other hypervisors in future. This area will never be used for real
HW, it is only used by software.

Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: config: enable pci host controller fdt driver

Upstream: no

When the virtual machine only uses the -kernel parameter
to boot the kernel but does not carry the -bios parameter,
the device information of the startup mainly relies on the
fdt to pass to the kernel instead of the acpi,

so we need to enable the pci host controller fdt driver
to solve the following call stack problem:

CPU 0 Unable to handle kernel paging request at virtual address ffff800000008064, era == 9000000001155e78, ra == 90000000015c124c

Call Trace:
[<9000000001155e78>] i8042_flush+0x88/0x1b8
[<90000000015c1248>] i8042_init+0x1c4/0x28c
[<9000000000220ccc>] do_one_initcall+0x6c/0x2a0
[<9000000001581054>] do_initcalls+0x114/0x164
[<900000000158128c>] kernel_init_freeable+0x174/0x1b0
[<9000000001565254>] kernel_init+0x24/0x120
[<9000000000221fa4>] ret_from_kernel_thread+0x8/0xa4

Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: KVM: Repair the restart pressure test virtual machine hang dead problem

Upstream: no

After the virtual restart, exintc does not correctly clear
the estat register and qemu does not clear the estat register.
Estat data sent from qemu is merged instead of overwritten on kvm.

In order to solve this problem,
the following changes were made to this patch:

1.exintc correctly clears zeros when the VM restarts
2. estat on the qemu side clears ESTAT correctly when the VM restarts
3. When the estat register data on the qemu side is sent to the kvm,
the data needs to be overwritten instead of merged

Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

mstsirkin committed:
virtgpu: don't reset on shutdown

commit 183a08715af1 ("LoongArch: Add ifdefs to fix LSX and LASX related warnings")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

It looks like GPUs are used after shutdown is invoked.
Thus, breaking virtio gpu in the shutdown callback is not a good idea -
guest hangs attempting to finish console drawing, with these warnings:

[   20.504464] WARNING: CPU: 0 PID: 568 at
drivers/gpu/drm/virtio/virtgpu_vq.c:358
virtio_gpu_queue_ctrl_sgs+0x236/0x290 [virtio_gpu]
[   20.505685] Modules linked in: nft_fib_inet nft_fib_ipv4
nft_fib_ipv6 nft_fib nft_reject_inet nf_reject_ipv4 nf_reject_ipv6
nft_reject nft_ct nft_chain_nat nf_nat nf_conntrack nf_defrag_ipv6
nf_defrag_ipv4 rfkill ip_set nf_tables nfnetlink vfat fat intel_rapl_msr
intel_rapl_common intel_uncore_frequency_common nfit libnvdimm kvm_intel
kvm rapl iTCO_wdt iTCO_vendor_support virtio_gpu virtio_dma_buf pcspkr
drm_shmem_helper i2c_i801 drm_kms_helper lpc_ich i2c_smbus virtio_balloon
joydev drm fuse xfs libcrc32c ahci libahci crct10dif_pclmul crc32_pclmul
crc32c_intel libata virtio_net ghash_clmulni_intel net_failover virtio_blk
failover serio_raw dm_mirror dm_region_hash dm_log dm_mod
[   20.511847] CPU: 0 PID: 568 Comm: kworker/0:3 Kdump: loaded Tainted: G
W         -------  ---  5.14.0-578.6675_1757216455.el9.x86_64 #1
[   20.513157] Hardware name: Red Hat KVM/RHEL, BIOS edk2-20241117-3.el9
11/17/2024
[ 20.513918] Workqueue: events drm_fb_helper_damage_work [drm_kms_helper]
[ 20.514626] RIP: 0010:virtio_gpu_queue_ctrl_sgs+0x236/0x290 [virtio_gpu]
[ 20.515332] Code: 00 00 48 85 c0 74 0c 48 8b 78 08 48 89 ee e8 51 50 00
00 65 ff 0d 42 e3 74 3f 0f 85 69 ff ff ff 0f 1f 44 00 00 e9 5f ff ff ff
<0f> 0b e9 3f ff ff ff 48 83 3c 24 00 74 0e 49 8b 7f 40 48 85 ff 74
[   20.517272] RSP: 0018:ff34f0a8c0787ad8 EFLAGS: 00010282
[   20.517820] RAX: 00000000fffffffb RBX: 0000000000000000
RCX: 0000000000000820
[   20.518565] RDX: 0000000000000000 RSI: ff34f0a8c0787be0
RDI: ff218bef03a26300
[   20.519308] RBP: ff218bef03a26300 R08: 0000000000000001
R09: ff218bef07224360
[   20.520059] R10: 0000000000008dc0 R11: 0000000000000002
R12: ff218bef02630028
[   20.520806] R13: ff218bef0263fb48 R14: ff218bef00cb8000
R15: ff218bef07224360
[   20.521555] FS:  0000000000000000(0000) GS:ff218bef7ba00000(0000)
knlGS:0000000000000000
[   20.522397] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   20.522996] CR2: 000055ac4f7871c0 CR3: 000000010b9f2002
CR4: 0000000000771ef0
[   20.523740] DR0: 0000000000000000 DR1: 0000000000000000
DR2: 0000000000000000
[   20.524477] DR3: 0000000000000000 DR6: 00000000fffe07f0
DR7: 0000000000000400
[   20.525223] PKRU: 55555554
[   20.525515] Call Trace:
[   20.525777]  <TASK>
[   20.526003]  ? show_trace_log_lvl+0x1c4/0x2df
[   20.526464]  ? show_trace_log_lvl+0x1c4/0x2df
[20.526925]  ? virtio_gpu_queue_fenced_ctrl_buffer+0x82/0x2c0 [virtio_gpu]
[   20.527643]  ? virtio_gpu_queue_ctrl_sgs+0x236/0x290 [virtio_gpu]
[   20.528282]  ? __warn+0x7e/0xd0
[   20.528621]  ? virtio_gpu_queue_ctrl_sgs+0x236/0x290 [virtio_gpu]
[   20.529256]  ? report_bug+0x100/0x140
[   20.529643]  ? handle_bug+0x3c/0x70
[   20.530010]  ? exc_invalid_op+0x14/0x70
[   20.530421]  ? asm_exc_invalid_op+0x16/0x20
[   20.530862]  ? virtio_gpu_queue_ctrl_sgs+0x236/0x290 [virtio_gpu]
[   20.531506]  ? virtio_gpu_queue_ctrl_sgs+0x174/0x290 [virtio_gpu]
[ 20.532148]  virtio_gpu_queue_fenced_ctrl_buffer+0x82/0x2c0 [virtio_gpu]
[   20.532843]  virtio_gpu_primary_plane_update+0x3e2/0x460 [virtio_gpu]
[ 20.533520]  drm_atomic_helper_commit_planes+0x108/0x320 [drm_kms_helper]
[   20.534233]  drm_atomic_helper_commit_tail+0x45/0x80 [drm_kms_helper]
[   20.534914]  commit_tail+0xd2/0x130 [drm_kms_helper]
[   20.535446]  drm_atomic_helper_commit+0x11b/0x140 [drm_kms_helper]
[   20.536097]  drm_atomic_commit+0xa4/0xe0 [drm]
[   20.536588]  ? __pfx___drm_printfn_info+0x10/0x10 [drm]
[   20.537162]  drm_atomic_helper_dirtyfb+0x192/0x270 [drm_kms_helper]
[20.537823]  drm_fbdev_shmem_helper_fb_dirty+0x43/0xa0 [drm_shmem_helper]
[   20.538536]  drm_fb_helper_damage_work+0x87/0x160 [drm_kms_helper]
[   20.539188]  process_one_work+0x194/0x380
[   20.539612]  worker_thread+0x2fe/0x410
[   20.540007]  ? __pfx_worker_thread+0x10/0x10
[   20.540456]  kthread+0xdd/0x100
[   20.540791]  ? __pfx_kthread+0x10/0x10
[   20.541190]  ret_from_fork+0x29/0x50
[   20.541566]  </TASK>
[   20.541802] ---[ end trace 0000000000000000 ]---

It looks like the shutdown is called in the middle of console drawing, so
we should either wait for it to finish, or let drm handle the shutdown.

This patch implements this second option:

Add an option for drivers to bypass the common break+reset handling.
As DRM is careful to flush/synchronize outstanding buffers, it looks like
GPU can just have a NOP there.

Reviewed-by: Eric Auger <eric.auger@redhat.com>
Tested-by: Eric Auger <eric.auger@redhat.com>
Fixes: 8bd2fa086a04 ("virtio: break and reset virtio devices on device_shutdown()")
Cc: Eric Auger <eauger@redhat.com>
Cc: Jocelyn Falempe <jfalempe@redhat.com>
Signed-off-by: Michael S. Tsirkin <mst@redhat.com>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>
Change-Id: I67ee2081bcb4efe30630bb00621603071cfcb466

bibo-mao committed:
LoongArch: KVM: Fix PMU pass-through issue if VM exits to host finally

commit 5add0dbbebd6 ("LoongArch: KVM: Fix PMU pass-through issue if VM exits to host finally")
Conflict: none
Backport-reason: Synchronize upstream linux loongarch kvm
patch to support loongarch virtualization.
Checkpatch: no, to be consistent with upstream commit.

In function kvm_pre_enter_guest(), it prepares to enter guest and check
whether there are pending signals or events. And it will not enter guest
if there are, PMU pass-through preparation for guest should be cancelled
and host should own PMU hardware.

Fixes: f4e40ea9f78f ("LoongArch: KVM: Add PMU support for guest")
Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: loongarch/configs: enable vfio mdev config

Upstream: no

enable vfio mdev config

Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

lixianglai committed:
LoongArch: driver/vfio: disable vfio pci intx

Upstream: no

Non-zero-node pci devices on multi-node devices
such as 3D6000 no longer support intx.
To avoid device pass-through failure caused by
detecting intx on the qemu side,
the vfio pci intx function is now disabled.

Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

Avenger-285714 commented:
/ok-to-test

lixianglai committed:
LoongArch: KVM: add virt extioi cpu encode support

Upstream: no

The extioi controller can send interrupts to only
four cpus and cannot send interrupts to other cpus.
This patch enables extioi to send interrupts to
a maximum of 256 cpus.

Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

opsiff commented on arch/loongarch/kernel/paravirt.c:
```diff
@@ -4,9 +4,9 @@
-#include <asm/paravirt.h>
 #include <linux/reboot.h>
 #include <linux/static_call.h>
+#include <asm/paravirt.h>
```
这里修改似乎没有什么意义

lixianglai replied:
恩。。是想对齐当前代码与其他国内社区代码的差距做的修改，这样可以保持各大社区虚拟化代码一致

opsiff replied:
> 恩。。是想对齐当前代码与其他国内社区代码的差距做的修改，这样可以保持各大社区虚拟化代码一致

https://github.com/deepin-community/kernel/pull/635/commits/3d59b16b50ac2bfd59848843b22648b56634ee10  这个补丁意义不明 需要修改格式规范 比如欧拉社区也不接收这种修改吧

lixianglai replied:
这个patch形成的原因确实比较复杂，有些代码片段是由于一个功能实现提交过多次（提交多次的原因是一些patch在没有提交到上游内核社区之前就已经合入过，在提交上游内核社区之后，发生了变化，又从上游回合过）导致有些修改会残留或者遗漏，因为代码已经合入仓库之前的提交无法改变了，所以零零碎碎的就合成了一个大patch，用来修复编译问题，并与其他社区保持一致

lixianglai replied:
另外还有一个原因就是，patch顺序的调整，冲突的解决过程中也会出问题

opsiff replied:
> 这个patch形成的原因确实比较复杂，有些代码片段是由于一个功能实现提交过多次（提交多次的原因是一些patch在没有提交到上游内核社区之前就已经合入过，在提交上游内核社区之后，发生了变化，又从上游回合过）导致有些修改会残留或者遗漏，因为代码已经合入仓库之前的提交无法改变了，所以零零碎碎的就合成了一个大patch，用来修复编译问题，并与其他社区保持一致

可以在commit msg中说明针对哪些补丁进行的修改，以及标题需要修改

lixianglai replied:
好的，那我修改一下commit msg

opsiff replied:
> 好的，那我修改一下commit msg

这个commit title前面需要加一下"Loongarch:"

lixianglai committed:
drm/virtio: implement virtio_gpu_shutdown

Upstream: no

Calling drm_dev_unplug() is the drm way to say the device
is gone and can not be accessed any more.

Cc: Michael S. Tsirkin <mst@redhat.com>
Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>
Reviewed-by: Eric Auger <eric.auger@redhat.com>
Tested-by: Eric Auger <eric.auger@redhat.com>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>
Change-Id: Ib23f2754d314ea61c4a341bcff9dcd05bdcea28d

gaosong-loongson committed:
LoongArch: fix migrate issue

Upstream: no

when read_fault, we shoul't set the page _PAGE_WRITE.
this may case nest write fault issue.

Signed-off-by: Song Gao <gaosong@loongson.cn>
Change-Id: I443ac32d152c43635c0bafbe3da9b9479f4e546d
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>

opsiff commented on arch/loongarch/mmu.c:
```diff
@@ -0,0 +1,995 @@
+#include <asm/tlb.h>
+#include <asm/kvm_mmu.h>
+
+static inline bool kvm_hugepage_capable(struct kvm_memory_slot *slot)
```
这个文件是不是多余了

lixianglai replied:
奥，提交错误我，修改一下

Avenger-285714 commented on drivers/irqchip/irq-loongson-eiointc.c:
```diff
@@ -176,13 +197,14 @@ static int eiointc_router_init(unsigned int cpu)
 	uint32_t data;
 	uint32_t node = cpu_to_eio_node(cpu);
 	int index = eiointc_index(node);
+	int cores = (cpu_has_hypervisor ? MAX_CORES_PER_EIO_NODE : CORES_PER_EIO_NODE);
```
drivers/irqchip/irq-loongson-eiointc.c:200:6: error: unused variable 'cores' [-Werror,-Wunused-variable]
  200 |         int cores = (cpu_has_hypervisor ? MAX_CORES_PER_EIO_NODE : CORES_PER_EIO_NODE);
      |             ^~~~~
1 error generated.

Avenger-285714 commented on arch/loongarch/kvm/intc/extioi.c:
```diff
@@ -0,0 +1,954 @@
+				gpa_t addr, int len, void *val)
+{
+	int index, ret = 0;
+	gpa_t offset;
```
arch/loongarch/kvm/intc/extioi.c:458:2: error: variable 'data' is used uninitialized whenever switch default is taken [-Werror,-Wsometimes-uninitialized]
  458 |         default:
      |         ^~~~~~~
arch/loongarch/kvm/intc/extioi.c:463:15: note: uninitialized use occurs here
  463 |         *(u8 *)val = data;
      |                      ^~~~
arch/loongarch/kvm/intc/extioi.c:429:10: note: initialize the variable 'data' to silence this warning
  429 |         u64 data;
      |                 ^
      |                  = 0
arch/loongarch/kvm/intc/extioi.c:502:2: error: variable 'data' is used uninitialized whenever switch default is taken [-Werror,-Wsometimes-uninitialized]
  502 |         default:
      |         ^~~~~~~
arch/loongarch/kvm/intc/extioi.c:507:16: note: uninitialized use occurs here
  507 |         *(u32 *)val = data;
      |                       ^~~~
arch/loongarch/kvm/intc/extioi.c:473:10: note: initialize the variable 'data' to silence this warning
  473 |         u64 data;
      |                 ^
      |                  = 0
arch/loongarch/kvm/intc/extioi.c:546:2: error: variable 'data' is used uninitialized whenever switch default is taken [-Werror,-Wsometimes-uninitialized]
  546 |         default:
      |         ^~~~~~~
arch/loongarch/kvm/intc/extioi.c:551:16: note: uninitialized use occurs here
  551 |         *(u64 *)val = data;
      |                       ^~~~
arch/loongarch/kvm/intc/extioi.c:517:10: note: initialize the variable 'data' to silence this warning
  517 |         u64 data;
      |                 ^
      |                  = 0
arch/loongarch/kvm/intc/extioi.c:582:2: error: variable 'ret' is used uninitialized whenever switch default is taken [-Werror,-Wsometimes-uninitialized]
  582 |         default:
      |         ^~~~~~~
arch/loongarch/kvm/intc/extioi.c:588:9: note: uninitialized use occurs here
  588 |         return ret;
      |                ^~~
arch/loongarch/kvm/intc/extioi.c:560:9: note: initialize the variable 'ret' to silence this warning
  560 |         int ret;
      |                ^
      |                 = 0
4 errors generated.

Avenger-285714 commented on drivers/iommu/loongarch_iommu.c:
```diff
@@ -0,0 +1,1854 @@
+	}
+}
+
+static int __init la_iommu_setup(char *str)
```
drivers/iommu/loongarch_iommu.c:1784:19: error: unused function 'la_iommu_setup' [-Werror,-Wunused-function]
 1784 | static int __init la_iommu_setup(char *str)
      |                   ^~~~~~~~~~~~~~
1 error generated.

lixianglai replied:
这个误报吧，我下面用了   __setup("loongarch_iommu=", la_iommu_setup);
有什么解决方法吗？
